src/java/org/apache/cassandra/tcm/transformations/AccordMarkStale.java

```diff
@@ -0,0 +1,119 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.tcm.transformations;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import accord.local.Node;
+import accord.topology.Shard;
+import org.apache.cassandra.io.util.DataInputPlus;
+import org.apache.cassandra.io.util.DataOutputPlus;
+import org.apache.cassandra.schema.KeyspaceMetadata;
+import org.apache.cassandra.schema.SchemaConstants;
+import org.apache.cassandra.service.accord.AccordTopology;
+import org.apache.cassandra.tcm.ClusterMetadata;
+import org.apache.cassandra.tcm.Transformation;
+import org.apache.cassandra.tcm.membership.NodeId;
+import org.apache.cassandra.tcm.sequences.LockedRanges;
+import org.apache.cassandra.tcm.serialization.AsymmetricMetadataSerializer;
+import org.apache.cassandra.tcm.serialization.Version;
+
+import static org.apache.cassandra.exceptions.ExceptionCode.INVALID;
+
+public class AccordMarkStale implements Transformation
+{
+
+    private final NodeId tcmId;
+
+    public AccordMarkStale(NodeId tcmId)
+    {
+        this.tcmId = tcmId;
+    }
+
+    @Override
+    public Kind kind()
+    {
+        return Kind.ACCORD_MARK_STALE;
+    }
+
+    @Override
+    public Result execute(ClusterMetadata prev)
+    {
+        if (!prev.directory.peerIds().contains(tcmId))
+            return new Rejected(INVALID, String.format("Can not mark %s stale it is not present in the directory.", tcmId));
+
+        Node.Id staleAccordId = AccordTopology.tcmIdToAccord(tcmId);
+
+        if (prev.staleReplicas.contains(staleAccordId))
+            return new Rejected(INVALID, String.format("Can not mark %s stale as it already is.", tcmId));
+
+        for (KeyspaceMetadata keyspace : prev.schema.getKeyspaces().without(SchemaConstants.REPLICATED_SYSTEM_KEYSPACE_NAMES))
+        {
+            List<AccordTopology.KeyspaceShard> shards = AccordTopology.KeyspaceShard.forKeyspace(keyspace, prev.placements, prev.directory);
+            
+            for (AccordTopology.KeyspaceShard shard : shards)
+            {
+                // We're trying to mark a node in this shard stale...
+                if (shard.nodes().contains(staleAccordId))
+                {
+                    int quorumSize = Shard.slowPathQuorumSize(shard.nodes().size());
+                    Set<Node.Id> nonStaleNodes = new HashSet<>(shard.nodes());
+                    nonStaleNodes.remove(staleAccordId);
+                    nonStaleNodes.removeAll(prev.staleReplicas.ids());
+
+                    // ...but reject the transformation if this would bring us below quorum.
+                    if (nonStaleNodes.size() < quorumSize)
+                        return new Rejected(INVALID, String.format("Can not mark %s stale as there are already too many stale nodes for range %s in keyspace '%s'.", 
```

Comment:
nit... rf=2 can't have a single down node... so your message isn't correct in that case (it doesn't "already" have too many stale nodes; it just would not be in quorum if we made this change)

Reply:
Yeah, the message needs to be corrected to indicate marking will *produce* a bad state.